services/socks5/service/service.go

```diff
@@ -0,0 +1,252 @@
+	"github.com/mysteriumnetwork/node/session"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+// GetProposal returns the proposal for SOCKS5 service
+func GetProposal(location location.Location) market.ServiceProposal {
+	marketLocation := market.Location{
+		Continent: location.Continent,
+		Country:   location.Country,
+		City:      location.City,
+
+		ASN:      location.ASN,
+		ISP:      location.ISP,
+		NodeType: location.NodeType,
+	}
+
+	return market.ServiceProposal{
+		ServiceType: socks5.ServiceType,
+		ServiceDefinition: socks5.ServiceDefinition{
+			Location:          marketLocation,
+			LocationOriginate: marketLocation,
+		},
+		PaymentMethodType: dto.PaymentMethodPerTime,
+		PaymentMethod:     dto.DefaultPaymentInfo,
+	}
+}
+
+// NATPinger defined Pinger interface for Provider
+type NATPinger interface {
+	BindServicePort(key string, port int)
+	Stop()
+	Valid() bool
+}
+
+// NATEventGetter allows us to fetch the last known NAT event
+type NATEventGetter interface {
+	LastEvent() *event.Event
+}
+
+// NewManager creates new instance of SOCKS5 service
+func NewManager(
+	options Options,
+	location location.ServiceLocationInfo,
+	natPort func(port int) (natRelease func()),
+	natEventGetter NATEventGetter,
+	natPinger NATPinger,
+) *Manager {
+	return &Manager{
+		natPort:        natPort,
+		natPinger:      natPinger,
+		natEventGetter: natEventGetter,
+		natPingerPorts: port.NewPool(),
+
+		servicePort: options.Port.Num(),
+		publicIP:    location.PubIP,
+		outboundIP:  location.OutIP,
+	}
+}
+
+// Manager represents an instance of SOCKS5 service
+type Manager struct {
+	natPort        func(int) (natPortRelease func())
+	natPinger      NATPinger
+	natPingerPorts port.ServicePortSupplier
+	natEventGetter NATEventGetter
+
+	servicePort int
+	publicIP    string
+	outboundIP  string
+
+	listener     net.Listener
+	listenerLock sync.Mutex
+	listenerStop chan struct{}
+}
+
+// ProvideConfig provides the config for consumer and handles new SOCKS5 connection.
+func (m *Manager) ProvideConfig(sessionRequestData json.RawMessage) (*session.ConfigParams, error) {
+	if m.servicePort == 0 {
+		return nil, errors.New("Service port not initialized")
+	}
+
+	sessionConfig := sock5_session.Response{
+		Port: m.servicePort,
+	}
+	sessionDestroy := func() {
+		// Do nothing
+	}
+
+	sessionRequest := sock5_session.Request{}
+	err := json.Unmarshal(sessionRequestData, &sessionRequest)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not unmarshal SOCKS5 session request")
+	}
+
+	natPingerEnabled := m.natPinger.Valid() && m.isBehindNAT() && m.portMappingFailed()
+	traversalParams, err := m.newTraversalParams(natPingerEnabled, sessionRequest)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create traversal params")
+	}
+
+	if natPingerEnabled {
+		log.Info().Msgf("NAT Pinger enabled, binding service port for proxy, key: %s", traversalParams.ProxyPortMappingKey)
```

Comment:
Will it work with NAT hole punching? SOCKS5 is running on TCP and hole punching is using UDP.

Reply:
Yes, as @anjmao noted, TCP hole punching mechanism is not supported in our implementation atm. For this we will need completely new TCP based pinger. I suspect that success rate for TCP hole punching will much lower. Idea is very similar at to one with UDP NAT, but now we need to control how we send TCP SYN packet to remote peer. 

Created related ticket #1680 

For now, I would would not attempt NAT hole punching in case upnp fails. We could even stop socks service if it is behind the nat and upnp failed.